tests/pytest/test_logic.py

```diff
@@ -0,0 +1,225 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from cocotb.types import Logic, Bit
+import pytest
+
+
+def test_logic_conversions():
+    Logic(0)
+    Logic(1)
+    Logic(True)
+    Logic(False)
+    Logic('0')
+    Logic('l')
+    Logic('L')
+    Logic('1')
+    Logic('h')
+    Logic('H')
+    Logic('x')
+    Logic('X')
+    Logic('w')
+    Logic('W')
+    Logic('u')
+    Logic('U')
+    Logic('z')
+    Logic('Z')
+    with pytest.raises(ValueError):
+        Logic('j')
+
+
+def test_bit_conversions():
+    Bit(0)
+    Bit(1)
+    Bit(False)
+    Bit(True)
+    Bit('0')
+    Bit('1')
+    with pytest.raises(ValueError):
+        _ = Bit(object())
+
+
+def test_bit_logic_conversions():
+    Logic(Bit(0))
+    Logic(Bit(1))
+    Bit(Logic(0))
+    Bit(Logic(1))
+    with pytest.raises(ValueError):
+        Bit(Logic('X'))
+    with pytest.raises(ValueError):
+        Bit(Logic('Z'))
+
+
+def test_logic_equality():
+    assert Logic(0) == Logic('0')
+    assert Logic(0) != Logic('X')
+    assert Logic(0) != object()
+
+
+def test_bit_equality():
+    assert Bit(0) == Bit(False)
+    assert Bit(1) != Bit('0')
+    assert Bit(1) != object()
+
+
+def test_logic_bit_equality():
+    assert Logic(0) != Bit(0)
+
+
+def test_logic_default_value():
+    assert Logic() == Logic('X')
+
+
+def test_bit_default_value():
+    assert Bit() == Bit('0')
+
+
+def test_logic_bool_conversions():
+    assert bool(Logic('1')) is True
+    assert bool(Logic('0')) is False
+    with pytest.raises(ValueError):
+        bool(Logic('X'))
+    with pytest.raises(ValueError):
+        bool(Logic('Z'))
+
+
+def test_bit_bool_conversions():
+    assert bool(Bit(1)) is True
+    assert bool(Bit(0)) is False
+
+
+def test_logic_str_conversions():
+    assert str(Logic('0')) == '0'
+    assert str(Logic('1')) == '1'
+    assert str(Logic('X')) == 'X'
+    assert str(Logic('Z')) == 'Z'
```

Comment:
```suggestion
    assert str(Logic('Z')) == 'Z'
    assert str(Logic('W')) == 'X'
    assert str(Logic('-')) == 'X'
    assert str(Logic('z')) == 'Z'
    assert str(Logic('x')) == 'X'
```
?

Reply:
I don't think we need to test that exactly if we can prove that `Logic('x') == Logic('-')`, etc. I don't have a test for that.

Reply:
Added tests for identity check between value constructed from different literals.